tests/sum_products.rs

```diff
@@ -1,11 +1,11 @@
-use numpy::{array, dot, einsum, inner, pyarray, PyArray1, PyArray2};
+use numpy::{array, dot, einsum, inner, pyarray, PyArray0, PyArray1, PyArray2};
 
 #[test]
 fn test_dot() {
     pyo3::Python::with_gil(|py| {
         let a = pyarray![py, [1, 0], [0, 1]];
         let b = pyarray![py, [4, 1], [2, 2]];
-        let c = dot(a, b).unwrap();
+        let c: &PyArray2<_> = dot(a, b).unwrap();
```

Comment:
To avoid needing to specify the result type for inference, do you think that there's a case for making a `Dot` trait?

Then we have something like `fn dot(x, T1, y: T2) -> <T1 as Dot::<T2>>::Result`, and can define all the return types based on input types.

Reply:
Is that possible for e.g. dynamically sized arrays? I think NumPy would still return `float64` if the `IxDyn` inputs where e.g. one-dimensional so that at least some cases would not be visible on the type level.

Reply:
(There is admittedly also my general preference for the ndarray API serving the same purpose, so that I somewhat inclined towards simplicity of implementation here compared to say usability.)

Reply:
((Finally, I am not even sure that the originally reported problem - getting `PyArray0` for `T=i32` but `f64` for `T=f64` - could be solved on the type-level without specialization?))

Reply:
> ((Finally, I am not even sure that the originally reported problem - getting PyArray0 for T=i32 but f64 for T=f64 - could be solved on the type-level without specialization?))

It appears that would not be a problem: One can `extract` a `T` from a zero-dimensional `numpy.ndarray` with element type `T`, i.e. returning a scalar should always work if the result is conceptually scalar. The bug report is for the other way around, i.e. one cannot `extract` a `PyArray0<T>` when a bare `T` is actually returned by the NumPy routine.

Reply:
Right, yep I understand now. Let's keep as you've designed 👍